spec/datadog/profiling/collectors/cpu_and_wall_time_worker_spec.rb

```diff
@@ -747,61 +807,83 @@
-        1.times(&new_object)
+          expect(after_allocations - before_allocations).to be 100
+        end
 
-        t1_can_run = Queue.new
-        t1_has_run = Queue.new
-        before_t1 = nil
-        after_t1 = nil
+        it 'returns different numbers of allocations for different threads' do
+          # To get the exact expected number of allocations, we run this once before so that Ruby can create and cache all
+          # it needs to
+          new_object = proc { Object.new }
+          1.times(&new_object)
 
-        background_t1 = Thread.new do
-          before_t1 = described_class._native_allocation_count
-          t1_can_run.pop
+          t1_can_run = Queue.new
+          t1_has_run = Queue.new
+          before_t1 = nil
+          after_t1 = nil
 
-          100.times(&new_object)
-          after_t1 = described_class._native_allocation_count
-          t1_has_run << true
+          background_t1 = Thread.new do
+            before_t1 = described_class._native_allocation_count
+            t1_can_run.pop
+
+            100.times(&new_object)
+            after_t1 = described_class._native_allocation_count
+            t1_has_run << true
+          end
+
+          before_allocations = described_class._native_allocation_count
+          t1_can_run << true
+          t1_has_run.pop
+          after_allocations = described_class._native_allocation_count
+
+          background_t1.join
+
+          # This test checks that even though we observed 100 allocations in a background thread t1, the counters for
+          # the current thread were not affected by this change
+
+          expect(after_t1 - before_t1).to be 100
+          expect(after_allocations - before_allocations).to be < 10
         end
+      end
 
-        before_allocations = described_class._native_allocation_count
-        t1_can_run << true
-        t1_has_run.pop
-        after_allocations = described_class._native_allocation_count
+      context 'when allocation profiling is disabled' do
+        let(:allocation_profiling_enabled) { false }
 
-        background_t1.join
+        it 'still returns nil after a bunch of allocations' do
+          # To get the exact expected number of allocations, we run this once before so that Ruby can create and cache all
+          # it needs to
+          new_object = proc { Object.new }
+          1.times(&new_object)
 
-        # This test checks that even though we observed 100 allocations in a background thread t1, the counters for
-        # the current thread were not affected by this change
+          before_allocations = described_class._native_allocation_count
+          100.times(&new_object)
+          after_allocations = described_class._native_allocation_count
 
-        expect(after_t1 - before_t1).to be 100
-        expect(after_allocations - before_allocations).to be < 10
+          expect(before_allocations).to be nil
+          expect(after_allocations).to be nil
+        end
```

Comment:
It does make sense to add a test for when this is disabled. But I don't quite get why we need to allocate anything at all as `_native_allocation_count` in this situation should return `nil` regardless of 0 or many objects being allocated. :eyes:

Reply:
Yeah this may be me being overly nitpicky about not relying on implementation details during tests.

The tests above validate that:
* `when CpuAndWallTimeWorker has not been started it is expected to be nil`
* `when CPUAndWallTimeWorker has been started and allocation profiling is enabled it returns the number of allocations between two calls of the mehod`

Treating the cpu and wall time worker as a black-box I can derive 2 logical possibilities from those tests:
* `nil` only returned when not started.
* `nil` always returned when allocation profiling disabled.

A test of the form: `when CPUAndWallTimeWorker has been started and allocation profiling is disabled it still returns nil after a bunch of allocations` completely disambiguates and covers the behaviour.

Doing just `when CPUAndWallTimeWorker has been started and allocation profiling is disabled it still returns nil` kinda leaves you wondering: ok but what happens after the first allocations are tracked? I feel like you'd need a separate `when CPUAndWallTimeWorker has been started and allocation profiling is enabled, it is set to 0 before we see allocations` or something.

Reply:
I see. To be fair, it wouldn't be unreasonable to have a test that says "when enabled it returns >= 0" ;P

---

That said, if the intention is to test "even after allocations, it'll still be nil", it seems very reasonable. But then.... why check the before allocations as well? Does this test need to be this thorough?

P.s.: I just spotted that the test isn't quite correct for another reason: we're not starting cpu and wall time worker >_> (oops)

Reply:
Ahah true that, also just spotted another weirdness, hierarchy-related. Test ends up being:

"_native_allocation_count when allocation profiling is enabled when CpuAndWallTimeWorker has been started when allocation profiling is disabled still returns nil after a bunch of allocations"

🙈 

I'll rework this.

Reply:
Yeah it's the usual problem of testing that things don't happen -- there may be multiple reasons for them not to happen, and sometimes it's not happening for the wrong reason ;)